### ISSUE TITLE
Nvidia-persistenced error

### ISSUE THREAD

**Issue by mkielak** [tkestack/gpu-manager#47]
Hi,

I try to run plugin on AWS with T4 card, 440.33.01 driver and CUDA 10.2. My GPU manager version is 1.10. GPU-admission is runing, GPU-manager pod is starting and running but I don't have vcuda-memory and vcuda-core in my node resources so my pod is stuck in FailedScheduling and error "1 Insufficient tencent.com/vcuda-core, 1 Insufficient tencent.com/vcuda-memory".
When I look at GPU-manager container logs I see some problem with nvidia-persistenced:

> rebuild ldcache     
> launch gpu manager  
> E1112 16:19:24.756540   24007 server.go:120] Can not start volume managerImpl, err /usr/local/nvidia/bin/nvidia-persistenced: bad magic number '[35 33 47 98]' in record at byte 0x0

When I enter into container and try to run /usr/local/nvidia/bin/nvidia-persistenced on my own it tries to access /usr/bin/nvidia-persistenced and stops. After creating symbolic link to /usr/bin there is no error when trying to run it manually but still error is the same.

**Comment by mYmNeo** [comment]
> bad magic number '[35 33 47 98]' in record at byte 0x0
This error means manager failed when we mirrors `nvidia-persistenced` from ld.cache. Could provide the log of gpu-manager    with the pattern `Mirror %s to %s`

**Comment by mkielak** [comment]
Here it is:

> I1113 12:06:00.086699 19692 volume.go:137] Find binaries: [/usr/local/nvidia/bin/nvidia-cuda-mps-control /usr/local/nvidia/bin/nvidia-cuda-mps-server /usr/local/nvidia/bin/nvidia-debugdump /usr/local/nvidia/bin/nvidia-persistenced /usr/local/nvidia/bin/nvidia-smi /usr/bin/gpu-client] 
I1113 12:06:00.088267 19692 volume.go:142] Find 32bit libraries: [/usr/local/nvidia/lib/libvdpau_nvidia.so.440.33.01 /usr/local/nvidia/lib/libnvidia-tls.so.440.33.01 /usr/local/nvidia/lib/libnvidia-ptxjitcompiler.so.440.33.01 /usr/local/nvidia/lib/libnvidia-opticalflow.so.440.33.01 /usr/local/nvidia/lib/libnvidia-opencl.so.440.33.01 /usr/local/nvidia/lib/libnvidia-ml.so.440.33.01 /usr/local/nvidia/lib/libnvidia-ifr.so.440.33.01 /usr/local/nvidia/lib/libnvidia-glsi.so.440.33.01 /usr/local/nvidia/lib/libnvidia-glcore.so.440.33.01 /usr/local/nvidia/lib/libnvidia-fbc.so.440.33.01 /usr/local/nvidia/lib/libnvidia-fatbinaryloader.so.440.33.01 /usr/local/nvidia/lib/libnvidia-encode.so.440.33.01 /usr/local/nvidia/lib/libnvidia-eglcore.so.440.33.01 /usr/local/nvidia/lib/libnvidia-compiler.so.440.33.01 /usr/local/nvidia/lib/libnvcuvid.so.440.33.01 /usr/local/nvidia/lib/libcuda.so.440.33.01 /usr/local/nvidia/lib/libOpenGL.so.0 /usr/local/nvidia/lib/libGLdispatch.so.0 /usr/local/nvidia/lib/libGLX_nvidia.so.440.33.01 /usr/local/nvidia/lib/libGLX.so.0 /usr/local/nvidia/lib/libGLESv2_nvidia.so.440.33.01 /usr/local/nvidia/lib/libGLESv2.so.2.1.0 /usr/local/nvidia/lib/libGLESv1_CM_nvidia.so.440.33.01 /usr/local/nvidia/lib/libGLESv1_CM.so.1.2.0 /usr/local/nvidia/lib/libGL.so.1.7.0 /usr/local/nvidia/lib/libEGL_nvidia.so.440.33.01 /usr/local/nvidia/lib/libEGL.so.1.1.0] 
I1113 12:06:00.088292 19692 volume.go:143] Find 64bit libraries: [/usr/local/nvidia/lib64/libvdpau_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-tls.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-ptxjitcompiler.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-opticalflow.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-opencl.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-ml.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-ifr.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-glsi.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-glcore.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-fbc.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-fatbinaryloader.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-encode.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-eglcore.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-egl-wayland.so.1.1.4 /usr/local/nvidia/lib64/libnvidia-compiler.so.440.33.01 /usr/local/nvidia/lib64/libnvcuvid.so.440.33.01 /usr/local/nvidia/lib64/libcuda.so.440.33.01 /usr/lib64/libcuda-control.so /usr/local/nvidia/lib64/libOpenGL.so.0 /usr/local/nvidia/lib64/libGLdispatch.so.0.0.0 /usr/local/nvidia/lib64/libGLX_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libGLX.so.0.0.0 /usr/local/nvidia/lib64/libGLESv2_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libGLESv2.so.2.1.0 /usr/local/nvidia/lib64/libGLESv1_CM_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libGLESv1_CM.so.1.2.0 /usr/local/nvidia/lib64/libGL.so.1.7.0 /usr/local/nvidia/lib64/libEGL_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libEGL.so.1.1.0]
I1113 12:06:00.089819 19692 volume.go:142] Find 32bit libraries: [/usr/local/nvidia/lib/libvdpau_nvidia.so.440.33.01 /usr/local/nvidia/lib/libnvidia-tls.so.440.33.01 /usr/local/nvidia/lib/libnvidia-ptxjitcompiler.so.440.33.01 /usr/local/nvidia/lib/libnvidia-opticalflow.so.440.33.01 /usr/local/nvidia/lib/libnvidia-opencl.so.440.33.01 /usr/local/nvidia/lib/libnvidia-ml.so.440.33.01 /usr/local/nvidia/lib/libnvidia-ifr.so.440.33.01 /usr/local/nvidia/lib/libnvidia-glsi.so.440.33.01 /usr/local/nvidia/lib/libnvidia-glcore.so.440.33.01 /usr/local/nvidia/lib/libnvidia-fbc.so.440.33.01 /usr/local/nvidia/lib/libnvidia-fatbinaryloader.so.440.33.01 /usr/local/nvidia/lib/libnvidia-encode.so.440.33.01 /usr/local/nvidia/lib/libnvidia-eglcore.so.440.33.01 /usr/local/nvidia/lib/libnvidia-compiler.so.440.33.01 /usr/local/nvidia/lib/libnvcuvid.so.440.33.01 /usr/local/nvidia/lib/libcuda.so.440.33.01 /usr/local/nvidia/lib/libOpenGL.so.0 /usr/local/nvidia/lib/libGLdispatch.so.0 /usr/local/nvidia/lib/libGLX_nvidia.so.440.33.01 /usr/local/nvidia/lib/libGLX.so.0 /usr/local/nvidia/lib/libGLESv2_nvidia.so.440.33.01 /usr/local/nvidia/lib/libGLESv2.so.2.1.0 /usr/local/nvidia/lib/libGLESv1_CM_nvidia.so.440.33.01 /usr/local/nvidia/lib/libGLESv1_CM.so.1.2.0 /usr/local/nvidia/lib/libGL.so.1.7.0 /usr/local/nvidia/lib/libEGL_nvidia.so.440.33.01 /usr/local/nvidia/lib/libEGL.so.1.1.0]
I1113 12:06:00.089831 19692 volume.go:143] Find 64bit libraries: [/usr/local/nvidia/lib64/libvdpau_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-tls.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-ptxjitcompiler.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-opticalflow.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-opencl.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-ml.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-ifr.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-glsi.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-glcore.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-fbc.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-fatbinaryloader.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-encode.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-eglcore.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-egl-wayland.so.1.1.4 /usr/local/nvidia/lib64/libnvidia-compiler.so.440.33.01 /usr/local/nvidia/lib64/libnvcuvid.so.440.33.01 /usr/local/nvidia/lib64/libcuda.so.440.33.01 /usr/local/nvidia/lib64/libOpenGL.so.0 /usr/local/nvidia/lib64/libGLdispatch.so.0.0.0 /usr/local/nvidia/lib64/libGLX_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libGLX.so.0.0.0 /usr/local/nvidia/lib64/libGLESv2_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libGLESv2.so.2.1.0 /usr/local/nvidia/lib64/libGLESv1_CM_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libGLESv1_CM.so.1.2.0 /usr/local/nvidia/lib64/libGL.so.1.7.0 /usr/local/nvidia/lib64/libEGL_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libEGL.so.1.1.0] 
I1113 12:06:00.090442 19692 volume.go:137] Find binaries: [/usr/local/nvidia/bin/nvidia-cuda-mps-control /usr/local/nvidia/bin/nvidia-cuda-mps-server /usr/local/nvidia/bin/nvidia-debugdump /usr/local/nvidia/bin/nvidia-persistenced /usr/local/nvidia/bin/nvidia-smi] 
I1113 12:06:00.090458 19692 volume.go:181] Mirror /usr/local/nvidia/bin/nvidia-cuda-mps-control to /etc/gpu-manager/vdriver/nvidia/bin 
I1113 12:06:00.090749 19692 volume.go:181] Mirror /usr/local/nvidia/bin/nvidia-cuda-mps-server to /etc/gpu-manager/vdriver/nvidia/bin 
I1113 12:06:00.090952 19692 volume.go:181] Mirror /usr/local/nvidia/bin/nvidia-debugdump to /etc/gpu-manager/vdriver/nvidia/bin 
I1113 12:06:00.091289 19692 volume.go:181] Mirror /usr/local/nvidia/bin/nvidia-persistenced to /etc/gpu-manager/vdriver/nvidia/bin E1113 12:06:00.091317 19692 server.go:122] Can not start volume managerImpl, err /usr/local/nvidia/bin/nvidia-persistenced: bad magic number '[35 33 47 98]' in record at byte 0x0

**Comment by mYmNeo** [comment]
> Here it is:
> 
> > I1113 12:06:00.086699 19692 volume.go:137] Find binaries: [/usr/local/nvidia/bin/nvidia-cuda-mps-control /usr/local/nvidia/bin/nvidia-cuda-mps-server /usr/local/nvidia/bin/nvidia-debugdump /usr/local/nvidia/bin/nvidia-persistenced /usr/local/nvidia/bin/nvidia-smi /usr/bin/gpu-client]
> > I1113 12:06:00.088267 19692 volume.go:142] Find 32bit libraries: [/usr/local/nvidia/lib/libvdpau_nvidia.so.440.33.01 /usr/local/nvidia/lib/libnvidia-tls.so.440.33.01 /usr/local/nvidia/lib/libnvidia-ptxjitcompiler.so.440.33.01 /usr/local/nvidia/lib/libnvidia-opticalflow.so.440.33.01 /usr/local/nvidia/lib/libnvidia-opencl.so.440.33.01 /usr/local/nvidia/lib/libnvidia-ml.so.440.33.01 /usr/local/nvidia/lib/libnvidia-ifr.so.440.33.01 /usr/local/nvidia/lib/libnvidia-glsi.so.440.33.01 /usr/local/nvidia/lib/libnvidia-glcore.so.440.33.01 /usr/local/nvidia/lib/libnvidia-fbc.so.440.33.01 /usr/local/nvidia/lib/libnvidia-fatbinaryloader.so.440.33.01 /usr/local/nvidia/lib/libnvidia-encode.so.440.33.01 /usr/local/nvidia/lib/libnvidia-eglcore.so.440.33.01 /usr/local/nvidia/lib/libnvidia-compiler.so.440.33.01 /usr/local/nvidia/lib/libnvcuvid.so.440.33.01 /usr/local/nvidia/lib/libcuda.so.440.33.01 /usr/local/nvidia/lib/libOpenGL.so.0 /usr/local/nvidia/lib/libGLdispatch.so.0 /usr/local/nvidia/lib/libGLX_nvidia.so.440.33.01 /usr/local/nvidia/lib/libGLX.so.0 /usr/local/nvidia/lib/libGLESv2_nvidia.so.440.33.01 /usr/local/nvidia/lib/libGLESv2.so.2.1.0 /usr/local/nvidia/lib/libGLESv1_CM_nvidia.so.440.33.01 /usr/local/nvidia/lib/libGLESv1_CM.so.1.2.0 /usr/local/nvidia/lib/libGL.so.1.7.0 /usr/local/nvidia/lib/libEGL_nvidia.so.440.33.01 /usr/local/nvidia/lib/libEGL.so.1.1.0]
> > I1113 12:06:00.088292 19692 volume.go:143] Find 64bit libraries: [/usr/local/nvidia/lib64/libvdpau_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-tls.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-ptxjitcompiler.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-opticalflow.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-opencl.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-ml.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-ifr.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-glsi.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-glcore.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-fbc.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-fatbinaryloader.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-encode.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-eglcore.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-egl-wayland.so.1.1.4 /usr/local/nvidia/lib64/libnvidia-compiler.so.440.33.01 /usr/local/nvidia/lib64/libnvcuvid.so.440.33.01 /usr/local/nvidia/lib64/libcuda.so.440.33.01 /usr/lib64/libcuda-control.so /usr/local/nvidia/lib64/libOpenGL.so.0 /usr/local/nvidia/lib64/libGLdispatch.so.0.0.0 /usr/local/nvidia/lib64/libGLX_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libGLX.so.0.0.0 /usr/local/nvidia/lib64/libGLESv2_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libGLESv2.so.2.1.0 /usr/local/nvidia/lib64/libGLESv1_CM_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libGLESv1_CM.so.1.2.0 /usr/local/nvidia/lib64/libGL.so.1.7.0 /usr/local/nvidia/lib64/libEGL_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libEGL.so.1.1.0]
> > I1113 12:06:00.089819 19692 volume.go:142] Find 32bit libraries: [/usr/local/nvidia/lib/libvdpau_nvidia.so.440.33.01 /usr/local/nvidia/lib/libnvidia-tls.so.440.33.01 /usr/local/nvidia/lib/libnvidia-ptxjitcompiler.so.440.33.01 /usr/local/nvidia/lib/libnvidia-opticalflow.so.440.33.01 /usr/local/nvidia/lib/libnvidia-opencl.so.440.33.01 /usr/local/nvidia/lib/libnvidia-ml.so.440.33.01 /usr/local/nvidia/lib/libnvidia-ifr.so.440.33.01 /usr/local/nvidia/lib/libnvidia-glsi.so.440.33.01 /usr/local/nvidia/lib/libnvidia-glcore.so.440.33.01 /usr/local/nvidia/lib/libnvidia-fbc.so.440.33.01 /usr/local/nvidia/lib/libnvidia-fatbinaryloader.so.440.33.01 /usr/local/nvidia/lib/libnvidia-encode.so.440.33.01 /usr/local/nvidia/lib/libnvidia-eglcore.so.440.33.01 /usr/local/nvidia/lib/libnvidia-compiler.so.440.33.01 /usr/local/nvidia/lib/libnvcuvid.so.440.33.01 /usr/local/nvidia/lib/libcuda.so.440.33.01 /usr/local/nvidia/lib/libOpenGL.so.0 /usr/local/nvidia/lib/libGLdispatch.so.0 /usr/local/nvidia/lib/libGLX_nvidia.so.440.33.01 /usr/local/nvidia/lib/libGLX.so.0 /usr/local/nvidia/lib/libGLESv2_nvidia.so.440.33.01 /usr/local/nvidia/lib/libGLESv2.so.2.1.0 /usr/local/nvidia/lib/libGLESv1_CM_nvidia.so.440.33.01 /usr/local/nvidia/lib/libGLESv1_CM.so.1.2.0 /usr/local/nvidia/lib/libGL.so.1.7.0 /usr/local/nvidia/lib/libEGL_nvidia.so.440.33.01 /usr/local/nvidia/lib/libEGL.so.1.1.0]
> > I1113 12:06:00.089831 19692 volume.go:143] Find 64bit libraries: [/usr/local/nvidia/lib64/libvdpau_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-tls.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-ptxjitcompiler.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-opticalflow.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-opencl.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-ml.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-ifr.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-glsi.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-glcore.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-fbc.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-fatbinaryloader.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-encode.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-eglcore.so.440.33.01 /usr/local/nvidia/lib64/libnvidia-egl-wayland.so.1.1.4 /usr/local/nvidia/lib64/libnvidia-compiler.so.440.33.01 /usr/local/nvidia/lib64/libnvcuvid.so.440.33.01 /usr/local/nvidia/lib64/libcuda.so.440.33.01 /usr/local/nvidia/lib64/libOpenGL.so.0 /usr/local/nvidia/lib64/libGLdispatch.so.0.0.0 /usr/local/nvidia/lib64/libGLX_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libGLX.so.0.0.0 /usr/local/nvidia/lib64/libGLESv2_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libGLESv2.so.2.1.0 /usr/local/nvidia/lib64/libGLESv1_CM_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libGLESv1_CM.so.1.2.0 /usr/local/nvidia/lib64/libGL.so.1.7.0 /usr/local/nvidia/lib64/libEGL_nvidia.so.440.33.01 /usr/local/nvidia/lib64/libEGL.so.1.1.0]
> > I1113 12:06:00.090442 19692 volume.go:137] Find binaries: [/usr/local/nvidia/bin/nvidia-cuda-mps-control /usr/local/nvidia/bin/nvidia-cuda-mps-server /usr/local/nvidia/bin/nvidia-debugdump /usr/local/nvidia/bin/nvidia-persistenced /usr/local/nvidia/bin/nvidia-smi]
> > I1113 12:06:00.090458 19692 volume.go:181] Mirror /usr/local/nvidia/bin/nvidia-cuda-mps-control to /etc/gpu-manager/vdriver/nvidia/bin
> > I1113 12:06:00.090749 19692 volume.go:181] Mirror /usr/local/nvidia/bin/nvidia-cuda-mps-server to /etc/gpu-manager/vdriver/nvidia/bin
> > I1113 12:06:00.090952 19692 volume.go:181] Mirror /usr/local/nvidia/bin/nvidia-debugdump to /etc/gpu-manager/vdriver/nvidia/bin
> > I1113 12:06:00.091289 19692 volume.go:181] Mirror /usr/local/nvidia/bin/nvidia-persistenced to /etc/gpu-manager/vdriver/nvidia/bin E1113 12:06:00.091317 19692 server.go:122] Can not start volume managerImpl, err /usr/local/nvidia/bin/nvidia-persistenced: bad magic number '[35 33 47 98]' in record at byte 0x0

The error is generated by elf read. It reports that your binary named `/usr/local/nvidia/bin/nvidia-persistenced` is a not elf format.

**Comment by mkielak** [comment]
I solved this issue. It was strange. I found out that size of nvidia-persistenced isn't the same as on host. It was caused by NVIDIA GLX. It has samples directory in /usr/share/doc/NVIDIA_GLX and there was placed another file called nvidia-persistenced which has smaller size. Proper nvidia-persistenced was copied but than it was overwritten. All of this happen inside copy-bin-lib.sh. After moving this samples folder somewhere I solved this issue.
